rclcpp/src/rclcpp/context.cpp

```diff
@@ -341,20 +346,35 @@ Context::shutdown(const std::string & reason)
   return true;
 }
 
-rclcpp::Context::OnShutdownCallback
-Context::on_shutdown(OnShutdownCallback callback)
+rclcpp::OnShutdownCallbackHandle::WeakPtr
+Context::on_shutdown(OnShutdownCallbackHandle::OnShutdownCallbackType callback)
+{
+  auto handle = std::make_shared<OnShutdownCallbackHandle>();
+  handle->callback = callback;
+  {
+    std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
+    on_shutdown_callbacks_.emplace(handle);
+  }
+  return handle;
+}
+
+void
+Context::remove_on_shutdown_callback(const OnShutdownCallbackHandle::WeakPtr callback_handle)
 {
-  on_shutdown_callbacks_.push_back(callback);
-  return callback;
+  auto callback_handle_shared_ptr = callback_handle.lock();
+  if (callback_handle_shared_ptr) {
+    std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
+    on_shutdown_callbacks_.erase(callback_handle_shared_ptr);
+  }
```

Comment:
Is it possible that between locking the weak ptr and acquiring the lock, this pointer could have been removed already?

I suppose it doesn't matter since `erase` won't care if it removed anything or not.

Reply:
> Is it possible that between locking the weak ptr and acquiring the lock, this pointer could have been removed already?

Yes. This scenario exists.

> I suppose it doesn't matter since erase won't care if it removed anything or not.

Yes. But I tend to modify it. `on_shutdown_callbacks_mutex_` should be locked at first.